[WIP] Initial attempt at reworking renaming

xarkes opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)


**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

This PR aims at reworking the renaming workflow.
Basically the shortcut `N` for "renaming" should handle *every* case, such as:
- Renaming a function
- Renaming a local variable
- Renaming a flag
- Adding a flag

***WIP***

- [ ] Update `FlagDialog` so it can manage renaming/editing an existing flag
- [ ] Make sure the dialog code is consistent with `CutterSeekable` when it comes to select a location
- [ ] Make sure renaming a local variable works

**Test plan (required)**

<!-- What steps should the reviewer take to test your pull request? Demonstrate that the code is solid. Example: The exact actions you made and their outcome. Add screenshots/videos if the pull request changes UI. This is your time to re-check that everything works and that you covered all the edge cases -->

For every case, it should be tried with a right click and with a direct call to the shortcut `N`.

- Rename a function
- Rename a function for which its first instruction contains a reference to an address e.g. `lea rdi, [0xc0fec0fe]`
- Rename an offset
- Rename a local variable
- Add a flag somewhere
- Delete a flag (make it empty)
- And so on...

<!-- **Code formatting**
Make sure you ran astyle on your code before making the PR. Check our contribution guidelines here: https://cutter.re/docs/code.html -->

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such). -->

This PR closes https://github.com/radareorg/cutter/issues/1698 and https://github.com/radareorg/cutter/issues/1211

ITAYC0HEN reviewed (changes requested):
I love it where I can N all the things :+1: 
As discussed, we should fix the dot issues with renaming in the disas. This will be a requirement for this PR.

In addition, a short observation. Taking this instruction for example:
```asm
0x00402bea      push    str.is_valid ; 0x4ca974
```

Clicking N on `str.is_valid` won't work -- that's the same one as before with the dot. With that, clicking N on "0x4ca974" in the auto comment does work which is very nice --- But, if I will manualy add "0x4ca974" as a comment in another instruction, it will won't work when I press N

```asm
0x00402bf4      add     esp, 0xc   ; Oh this is related to offset 0x4ca974   nice
```

xarkes commented:
Latest commit addresses your comment.
https://github.com/radareorg/cutter/pull/2473 is waiting for review.

xarkes commented:
Latest commit addresses your comment and https://github.com/radareorg/cutter/pull/2473 is waiting to be merged.

karliss commented on src/menus/DisassemblyContextMenu.cpp:
```diff
@@ -410,6 +361,105 @@ void DisassemblyContextMenu::setCanCopy(bool enabled)
+        }
+    }
+
+    if (!tuh) {
```
Slightly better way for dealing with situations where we weren't able to guess what exactly user wanted to rename would be task user to clarify. This would also reduce situations where it's impossible to rename something (without manually using r2 command) because the selection logic isn't good enough. Here is my proposal:

* When there is exact match of offset or name behave the same as now
* When there is no match but single thing that could be renamed rename that, in reality  this would mean empty things array since it's possible to add label at current offset
* Otherwise create menu with item for each possibility
    * When using shortcut (N) show this menu
    * When opening the regular context menu add it as submenu in the place that would normally have action for renaming single thing

karliss replied:
Do you disagree with this as concept or just don't want implementing right now? It is fine not doing it within this PR.

xarkes replied:
In a general fashion, I would tend to disagree with this approach as most of the time I prefer a tool to be reactive rather than opening yet another dialog box and requiring more interactivity.
Now, if this approach is desired for Cutter, I think we can discuss about it in a separate issue.
The purpose of this PR was to move every action under the `n` shortcut so that a user can rename anything that is being selected, and if it fallbacks to the default behavior or something the user did not want to rename, it just means that Cutter cannot rename the current selection.

karliss replied:
Yes I somewhat agree that in ideal world tool should know exactly what you intended to do, but fully achieving that is impossible without major changes to r2 and all the disassemblers. If you want a robust tool you can't only make the UI for easy cases ignoring the reality and hard cases. In the hard case telling "sorry Cutter can't rename this variable because developer didn't want you show one more dialog, go use different tool" is worse than one extra step in the hard case with no effect on the easy cases.

karliss replied:
Closing this for now to make it easier seeing what remains within this PR.

karliss commented on src/dialogs/FlagDialog.cpp:
```diff
@@ -12,19 +12,32 @@ FlagDialog::FlagDialog(RVA offset, QWidget *parent) :
 void FlagDialog::on_buttonBox_accepted()
 {
-    QString name = ui->nameEdit->text();
+    QString name = ui->nameEdit->text().replace(' ', '_');
```
I am not fan of custom logic for deciding what are allowed symbols all over the place. Especially if you ask using one name but it renames it something else. Such things would better be handled at r2 level.

xarkes replied:
I totally agree, I just found it convenient as a temporary fix, I can remove that.

xarkes commented on src/menus/DisassemblyContextMenu.cpp:
```diff
@@ -1011,11 +988,6 @@ void DisassemblyContextMenu::on_actionDeleteComment_triggered()
     Core()->delComment(offset);
 }
 
-void DisassemblyContextMenu::on_actionDeleteFlag_triggered()
```
Ah! Because I like the fact that renaming to empty string deletes the flag. However I think I forgot to test such behavior.

karliss replied:
I don't think it's a good reason for removing action to do it directly. You can keep and use the renaming to empty string behavior you like it, but having a dedicated button that does what it says is a lot more discoverable. Just because you can turn left by turning right 270 degrees doesn't  mean that having a steering wheels which can be turned both ways are useless and should be banned.

xarkes replied:
Correct, but my point of view is that the context menu is already very heavy, so I preferred to remove the action.
I can revert that change.

karliss replied:
Please do. I don't think space savings are worth the potential confusion due to not having an obvious way of removing flag.

If you want to save space, better make sure it is only shown when there is flag at current address. Also undefine function could be hidden when there is no function at current address.

xarkes replied:
Will do.

xarkes commented on src/dialogs/FlagDialog.h:
```diff
@@ -25,6 +25,7 @@ private slots:
 private:
     std::unique_ptr<Ui::FlagDialog> ui;
     RVA offset;
+    RFlagItem* flag;
```
So what's the fix? Copy the flag data in a temporary structure?

karliss replied:
Yes, looks like you don't need anything more than offset and name.

karliss commented on src/dialogs/FlagDialog.cpp:
```diff
@@ -40,17 +46,16 @@ void FlagDialog::buttonBoxAccepted()
 
     if (name.isEmpty()) {
-        if (flag) {
+        if (flagOffset) {
```
How about using RVA_INVALID instead of 0. 0 there might be something meaningful you want to flag at 0 for firmware or other raw binaries.

karliss commented:
Renaming near function header behaves quite bad:
* Click on 397 -> add flag at 0x18d , I understand how this happens but it's not great
* Click on fcn -> rename fcn.00006390
* Click on 00006390 -> add flag at  0x33, no idea how it got 0x33
* click on `var_1h` -> rename fcn.00006390, clicking on local variable names in other places work, it doesn't work only in variable list at top of function
* click on push -> rename fcn.00006390
* click on rbp -> rename flag rsp at 0x178000 ?? Probably shouldn't allow renaming flags in register namespace.
![Screenshot from 2020-11-18 18-44-21](https://user-images.githubusercontent.com/7101031/99560104-2c68ce80-29ce-11eb-9727-75dfdb5d6f8b.png)

xarkes commented:
I addressed some of your comments.
The one I could not fix is the "click on RSP" -> "renamed flag at". The reason for this is that radare2 adds flags for registers and in the case of this binary for some reason radare2 sets the RSP flag at `0x178000`. So given the renaming logic, I think it makes sense. I don't know what would be the proper way to handle this.

karliss commented on src/menus/DisassemblyContextMenu.cpp:
```diff
@@ -803,77 +816,38 @@ void DisassemblyContextMenu::on_actionAnalyzeFunction_triggered()
-        } else if (type == ThingUsedHere::Type::Function) {
-            Core()->triggerFunctionRenamed(thingUsedHere.offset, newName);
+    } else if (doRenameAction == RENAME_ADD_FLAG) {
+        FlagDialog dialog(doRenameInfo.addr, this->parentWidget());
```
Adding flag at current line seems to be broken. I think it worked in earlier version of this PR.

karliss reviewed (changes requested):
One more thing I forgot to mention before, but currently verify due to flag adding being broken. The rename action wasn't updated after adding new flag. So after pressing N first time to add a flag, pressing it second time still showed add flag dialog instead of rename dialog. Should be easy to fix.

xarkes commented:
So I added back the "Delete Flag" menu item.
I improved the FlagDialog so it actually returns if the input was confirmed or not (as any Dialog should do) and I am using this to refresh the renaming menu, in case the user doesn't click somewhere before trying to renaming again the current input.
Also I fixed the Dialog windows so their parent is the MainWindow and now they all appear at the same position (which is the center of the screen for my setup).
I fixed the small regression you mentioned as well.

Although it's not perfect, I like the current iteration of this PR, let me know what you think.

karliss commented on src/dialogs/FlagDialog.cpp:
```diff
@@ -62,9 +62,11 @@ void FlagDialog::buttonBoxAccepted()
 void FlagDialog::buttonBoxRejected()
 {
     close();
+    this->setResult(QDialog::Rejected);
```
Common pattern for using DialogButton box is as follows:
* connect `&QDialogButtonBox::accepted` to `&QDialog::accept`
* connect `&QDialogButtonBox::rejected` to `&QDialog::reject`
* either connect `&QDialog::accepted`  to the method which contains the logic you want do in dialog or check the dialog result in place you invoked it

First two steps can even be done in UI file.

karliss reviewed (approved):
ButtonBox signal handling cleanup is optional.

ITAYC0HEN reviewed (approved):
<3

ITAYC0HEN reviewed (changes requested):
update docs

xarkes commented:
Fixed the doc and pushed on top of master.

ITAYC0HEN commented on src/menus/DisassemblyContextMenu.cpp:
```diff
@@ -803,77 +822,37 @@ void DisassemblyContextMenu::on_actionAnalyzeFunction_triggered()
-        RVA offset = thingUsedHere.offset;
-        title = tr("Add flag at %1").arg(RAddressString(offset));
-        inputValue = "label." + QString::number(offset, 16);
+    } else if (doRenameAction == RENAME_DO_NOTHING) {
+        // Do nothing
     } else {
```
Wouldn't it be better to do only `else if (doRenameAction != RENAME_DO_NOTHING)` ?

xarkes replied:
In that way it is explicit that we want to do nothing when this specific action is chosen.
With your suggestion the handling of `RENAME_DO_NOTHING` would not be visible and someone might wonder later why it is not handled. I prefer it to be explicit.
